docs/architecture/adr-039-add-support-for-wasm-light-clients.md

```diff
@@ -0,0 +1,107 @@
+# ADR 39: Add support for Wasm based light client
+
+## Changelog
+
+- 26/11/2020: Initial Draft
+
+## Status
+
+Draft
+
+## Abstract
+
+Currently in the SDK, light clients are part of Cosmos SDK codebase. This makes upgrading existing IBC light client or add
+support for new light client is multi step process which is time-consuming.
+
+To remedy this, we are proposing a WASM VM to host light client bytecode, which allows easier upgrading of
+existing IBC light clients as well as adding support for new IBC light clients without requiring a code release and corresponding
+hard-fork event.
+
+## Context
+Currently in the SDK, light clients are defined as part of the codebase and are implemented as submodules under
+`/x/ibc/light-clients/`.
+
+Adding support for new light client or update an existing light client in the event of security
+issue or consensus update is multi-step process:
+
+1. Light client integration with SDK: IBC light clients are defined as part of the codebase and are implemented as
+submodules under `/x/ibc/light-clients/`. To add support for new light client or update an existing light client in the
+event of security issue or consensus update, we need to modify the codebase and integrate it in numerous places.
+
+2. Governance voting: Adding new light client implementations require governance support and is expensive: This is
+not ideal as chain governance is gatekeeper for new light client implementations getting added. If a small community
+want support for light client X, they may not be able to convince governance to support it.
+
+3. Validator upgrade: After governance voting succeeds, validators need to upgrade their nodes in order to enable new
+IBC light client implementation. This is both time consuming and error prone.
+   
+Another problem stemming from above process is, if a chain want to upgrade its own consensus, it will need to convince every chain
```

Comment:
```suggestion
Another problem stemming from the above process is that if a chain want to upgrade its own consensus, it will need to convince every chain
```